File: src/main/scala/com/algorithmlx/inscribers/api/client/InscribersButtonAPI.scala
package com.algorithmlx.inscribers.api.client

import com.algorithmlx.inscribers.api.client.InscribersButtonAPI.getActivated
import com.mojang.blaze3d.matrix.MatrixStack
import com.mojang.blaze3d.systems.RenderSystem
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.widget.button.Button
import net.minecraft.client.gui.widget.button.Button.{IPressable, ITooltip}
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraft.util.text.{ITextComponent, StringTextComponent}

class InscribersButtonAPI private(
  private val buttonX: Int,
  private val buttonY: Int,
  private val buttonWidth: Int,
  private val buttonHeight: Int,
  private val id: Int,
  private val text: ITextComponent,
  onPress: IPressable,
  tooltip: ITooltip,
  private val reloc: ResourceLocation,
  private val stack: ItemStack,
  private val stackedTexture: Boolean,
  private val enableDoubleButton: Boolean
) extends Button(buttonX, buttonY, buttonWidth, buttonHeight, text, onPress, tooltip) {

  def this(x: Int, y: Int, width: Int, height: Int, id: Int, text: ITextComponent, onPress: IPressable, tooltip: ITooltip, reloc: ResourceLocation, enableDoubleButton: Boolean) = {
    this(x, y, width, height, id, text, onPress, tooltip, reloc, ItemStack.EMPTY, false, enableDoubleButton)
  }

  def this(x: Int, y: Int, width: Int, height: Int, id: Int, text: ITextComponent, onPress: IPressable, reloc: ResourceLocation) = {
    this(x, y, width, height, id, text, onPress, (_: Button, _: MatrixStack, _: Int, _: Int) => {}, reloc, true)
  }

  def this(x: Int, y: Int, width: Int, height: Int, id: Int, text: ITextComponent, onPress: IPressable, tooltip: ITooltip, stack: ItemStack, enableDoubleButton: Boolean) = {
    this(x, y, width, height, id, text, onPress, tooltip, null, stack, true, enableDoubleButton)
  }

  def this(x: Int, y: Int, width: Int, height: Int, id: Int, text: ITextComponent, onPress: IPressable, stack: ItemStack, enableDoubleButton: Boolean) = {
    this(x, y, width, height, id, text, onPress, (_: Button, _: MatrixStack, _: Int, _: Int) => {}, stack, enableDoubleButton)
  }

  override def render(poseStack : MatrixStack, mouseX : Int, mouseY : Int, partialTick : Float): Unit = {
    val mc = Minecraft.getInstance()
    val font = mc.font

    poseStack.pushPose()
    if (this.text != null)
      font.draw(poseStack, this.text, this.buttonX * 1F, (this.buttonY + this.buttonHeight) / 4F, 0xFFFFFF)

    if (!this.stackedTexture) {
      val textureW = this.buttonWidth / 2
      val textureH = this.buttonHeight / 2
      Minecraft.getInstance().getTextureManager.bind(this.reloc)
      if (this.enableDoubleButton) {
        if (!getActivated(id)) {
          if (!this.isCursorAtButton(mouseX, mouseY)) {
            blit(poseStack, textureW, textureH, 0, 0, this.buttonWidth, this.buttonHeight)
          } else {
            blit(poseStack, textureW, textureH, 0, 8, this.buttonWidth, this.buttonHeight)
          }
        } else {
          if (!this.isCursorAtButton(mouseX, mouseY)) {
            blit(poseStack, textureW, textureH, 8, 0, this.buttonWidth, this.buttonHeight)
          } else {
            blit(poseStack, textureW, textureH, 8, 8, this.buttonWidth, this.buttonHeight)
          }
        }
      }
    } else {
      this.renderItem(poseStack, this.stack, this.buttonX, this.buttonY)
    }

    poseStack.popPose()
  }

  protected def renderItem(poseStack: MatrixStack, stack: ItemStack, x: Int, y: Int): Unit = {
    RenderSystem.pushMatrix()
    RenderSystem.multMatrix(poseStack.last().pose())
    poseStack.pushPose()
    Minecraft.getInstance().getItemRenderer.renderAndDecorateItem(stack, x, y)
    poseStack.popPose()
    RenderSystem.popMatrix()
  }

  def isCursorAtButton(cursorX: Int, cursorY: Int): Boolean = cursorX >= buttonX && cursorY >= buttonY && cursorX <= buttonX + buttonWidth&& cursorY <= buttonY + buttonHeight
}

object InscribersButtonAPI {
  private var activated = false
  private var id = 0

  def getActivated: Boolean = getActivated(getId)

  def setActivated(value: Boolean): Unit = this.setActivated(value, getId)

  def getActivated(id: Int): Boolean = this.id == id && this.activated

  def setActivated(value: Boolean, id: Int): Unit = {
    this.activated = value
    this.setId(id)
  }

  def getId: Int = this.id

  def setId(id: Int): Unit = {
    this.id = id
  }
}

File: src/main/java/com/algorithmlx/inscribers/api/client/InscribersButtonAPI.kt
package com.algorithmlx.inscribers.api.client

import com.algorithmlx.inscribers.api.drawString
import com.algorithmlx.inscribers.api.mouseAtPosition
import com.algorithmlx.inscribers.api.mc
import com.algorithmlx.inscribers.api.renderItem
import com.mojang.blaze3d.matrix.MatrixStack
import net.minecraft.client.gui.widget.button.Button
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraft.util.text.ITextComponent
import net.minecraft.util.text.StringTextComponent
import net.minecraft.util.text.TextFormatting

open class InscribersButtonAPI(
    private val buttonX: Int,
    private val buttonY: Int,
    private val buttonWidth: Int,
    private val buttonHeight: Int,
    private val id: Int,
    private val text: ITextComponent = StringTextComponent(""),
    onPress: IPressable,
    tooltip: ITooltip = ITooltip { _, _, _, _ -> },
    private val texture: ResourceLocation? = null,
    private val stack: ItemStack =  ItemStack.EMPTY,
    private val stackedTexture: Boolean = false,
    private val enableDoubleButton: Boolean = true
): Button(buttonX, buttonY, buttonWidth, buttonHeight, text, onPress, tooltip) {
    override fun render(pPoseStack: MatrixStack, pMouseX: Int, pMouseY: Int, pPartialTicks: Float) {
        pPoseStack.pushPose()
        if (this.text != StringTextComponent(""))
            pPoseStack.drawString(this.text, this.buttonX * 1F, (this.buttonY + this.buttonHeight) / 4F)

        if (!this.stackedTexture && this.texture != null) {
            val textureW = this.buttonWidth / 2
            val textureH = this.buttonHeight / 2
            mc.textureManager.bind(this.texture)
            if (this.enableDoubleButton) {
                if (!getActivated(this.id)) {
                    if (mouseAtPosition(pMouseX, pMouseY, pMouseX, pMouseY, pMouseX, pMouseY))
                        blit(pPoseStack, textureW, textureH, 0, 0, this.buttonWidth, this.buttonHeight)
                    else blit(pPoseStack, textureW, textureH, 0, 8, this.buttonWidth, buttonHeight)
                } else {
                    if (!mouseAtPosition(pMouseX, pMouseY, pMouseX, pMouseY, pMouseX, pMouseY))
                        blit(pPoseStack, textureW, textureH, 8, 0, this.buttonWidth, this.buttonHeight)
                    else blit(pPoseStack, textureW, textureH, 8, 8, this.buttonWidth, this.buttonHeight)
                }
            } else {
                pPoseStack.renderItem(this.stack, this.buttonX, this.buttonY)
            }
        }
    }

    companion object {
        private var activated = false
        var id = 0

        fun getActivated(): Boolean = getActivated(id)

        fun setActivated(value: Boolean) {
            this.setActivated(value, this.id)
        }

        fun getActivated(id: Int): Boolean = this.id == id  && this.activated

        fun setActivated(value: Boolean, id: Int) {
            this.activated = value
            this.id = id
        }
    }
}